File: tree/Construct_all_possible_BSTs_for_keys_1_to_N/construct2.cpp
#include <iostream> 
#include <vector> 
using namespace std; 

struct node {
    int data ; 
    struct node* right ; 
    struct node* left ; 
} ; 

struct node* newNode(int data) {
    struct node* temp = new struct node ; 
    temp->data = data ; 
    temp->left = NULL ; 
    temp->right = NULL ; 

    return temp ; 
}

void preorder(struct node* root){
    if(root == NULL){
        return ; 
    }

    cout<<root->data<<" "; 
    preorder(root->left); 
    preorder(root->right);
}


vector<struct node*> constructTree (int start, int end){
    vector<struct node*> lst ; 
    if(start > end){
        lst.push_back(NULL);
        return lst ; 
    }

    for(int i = start; i < end; i++){
        vector<struct node*> leftsubtree = constructTree(start, i-1) ; 
        vector<struct node*> rightsubtree = constructTree(i+1, end) ; 
        for(int j = 0 ; j < leftsubtree.size() ; j++){
            struct node* left = leftsubtree[j] ; 
            for(int k = 0 ; k < rightsubtree.size() ; k++){ 
                struct node* right = rightsubtree[k] ; 
                struct node* node = newNode(i) ;

                node->left = left ; 
                node->right = right ; 

                lst.push_back(node) ;  
            }
        }
    }

    return lst ; 
}

int main(){
    vector<struct node*> lst = constructTree(1, 5) ; 
    for(int i = 0 ; i < lst.size() ; i++){
        cout<<"Preorder traversal : "; 
        preorder(lst[i]);
        cout<<"\n" ; 
    }

    return 0 ; 
}

File: tree/Construct_BST_from_given_preorder/Method1/construct2.js
// A O(n^2) Python3 program for
// construction of BST from preorder traversal
 
// A binary tree node
 
 
class Node{
    // A constructor to create a new node
    constructor(data){
        this.data = data
        this.left = null
        this.right = null
    }
}
 
 
// constructTreeUtil.preIndex is a static variable of
// function constructTreeUtil
 
// Function to get the value of static variable
// constructTreeUtil.preIndex
function getPreIndex(){
    return constructTreeUtil.preIndex
}
 
// Function to increment the value of static variable
// constructTreeUtil.preIndex
 
 
function incrementPreIndex(){
    constructTreeUtil.preIndex += 1
}
 
// A recurseive function to construct Full from pre[].
// preIndex is used to keep track of index in pre[[].
 
 
function constructTreeUtil(pre, low, high){
 
        // Base Case
    if(low > high)
        return null
 
    // The first node in preorder traversal is root. So take
    // the node at preIndex from pre[] and make it root,
    // and increment preIndex
    let root = new Node(pre[getPreIndex()])
    incrementPreIndex()
 
    // If the current subarray has onlye one element,
    // no need to recur
    if(low == high)
        return root
 
    let r_root = -1
 
    // Search for the first element greater than root
    for(let i=low;i<high+1;i++){
        if (pre[i] > root.data){
            r_root = i
            break
        }
    }
 
    // If no elements are greater than the current root,
    // all elements are left children
    // so assign root appropriately
    if(r_root == -1)
        r_root = getPreIndex() + (high - low)
 
    // Use the index of element found in preorder to divide
    // preorder array in two parts. Left subtree and right
    // subtree
    root.left = constructTreeUtil(pre, getPreIndex(), r_root-1)
 
    root.right = constructTreeUtil(pre, r_root, high)
 
    return root
}
 
// The main function to construct BST from given preorder
// traversal. This function mailny uses constructTreeUtil()
 
 
function constructTree(pre){
    let size = pre.length
    constructTreeUtil.preIndex = 0
    return constructTreeUtil(pre, 0, size-1)
}
 
function printInorder(root){
    if(root == null)
        return
    printInorder(root.left)
    console.log(root.data,' ')
    printInorder(root.right)
}
 
 
// Driver code
let pre = [10, 5, 1, 7, 40, 50]
 
let root = constructTree(pre)
console.log("Inorder traversal of the constructed tree:","</br>")
printInorder(root)